receiver/receivertest/contract_checker.go

```diff
@@ -0,0 +1,382 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package receivertest // import "go.opentelemetry.io/collector/receiver/receivertest"
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/rand"
+	"sync"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/component/componenttest"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/consumer/consumererror"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	"go.opentelemetry.io/collector/receiver"
+)
+
+// UniqueIDAttrName is the attribute name that is used in log records/spans/datapoints as the unique identifier.
+const UniqueIDAttrName = "test_id"
+
+// UniqueIDAttrDataType is the value type of the UniqueIDAttrName.
+type UniqueIDAttrDataType int64
+
+type Generator interface {
+	// Generate must generate and send at least one data element (span, log record or metric data point)
+	// to the receiver and return a copy of generated element ids.
+	// The generated data must contain uniquely identifiable elements, each with a
+	// different value of attribute named UniqueIDAttrName.
+	// CreateOneLogWithID() can be used a helper to create such logs.
+	Generate() IDSet
+}
+
+type CheckConsumeContractParams struct {
+	T *testing.T
+	// Factory that allows to create a receiver.
+	Factory receiver.Factory
+	// Config of the receiver to use.
+	Config component.Config
+	// Generator that can send data to the receiver.
+	Generator Generator
+	// GenerateCount specifies the number of times to call the generator.Generate().
+	GenerateCount int
+	// ConsumeDecisionFunc defines the decision function to use when the receiver calls Consume() func
+	// the next consumer. ConsumeDecisionFunc defines the testing scenario (i.e. to test for
+	// success case or for error case or a mix of both). See for example RandomErrorsConsumeDecision.
+	ConsumeDecisionFunc consumeDecisionFunc
+}
+
+// CheckConsumeContract checks the contract between the receiver and its next consumer. For the contract
+// description see ../doc.go. The checker will detect violations of contract on success, on permanent
+// and non-permanent errors based on decision-making done by consumeDecisionFunc.
+// It is advised to run CheckConsumeContract with a variety of decision-making functions.
+func CheckConsumeContract(params CheckConsumeContractParams) {
+	consumer := &mockConsumer{t: params.T, consumeDecisionFunc: params.ConsumeDecisionFunc}
+	ctx := context.Background()
+
+	// Create and start the receiver.
+	receiver, err := params.Factory.CreateLogsReceiver(ctx, NewNopCreateSettings(), params.Config, consumer)
```

Comment:
This `CheckConsumeContract` function is applied to logs only, so we probably need to rename it or specify a signal type in `CheckConsumeContractParams`?